x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/index/fielddata/ShapeValues.java

```diff
@@ -0,0 +1,259 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.spatial.index.fielddata;
+
+import org.apache.lucene.document.ShapeField;
+import org.apache.lucene.geo.Component2D;
+import org.apache.lucene.util.BytesRef;
+import org.elasticsearch.geometry.Geometry;
+import org.elasticsearch.geometry.utils.GeographyValidator;
+import org.elasticsearch.geometry.utils.WellKnownText;
+import org.elasticsearch.index.mapper.ShapeIndexer;
+import org.elasticsearch.search.aggregations.support.ValuesSourceType;
+import org.elasticsearch.xcontent.ToXContentFragment;
+import org.elasticsearch.xcontent.XContentBuilder;
+import org.elasticsearch.xpack.spatial.index.mapper.BinaryShapeDocValuesField;
+
+import java.io.IOException;
+import java.text.ParseException;
+import java.util.function.Supplier;
+
+/**
+ * A stateful lightweight per document geo values.
+ * To iterate over values in a document use the following pattern:
+ * <pre>
+ *   MultiGeoValues values = ..;
+ *   // for each docID
+ *   if (values.advanceExact(docId)) {
+ *     GeoValue value = values.value()
+ *     final int numValues = values.count();
+ *     // process value
+ *   }
+ * </pre>
+ *
+ * There is just one value for one document.
+ */
+public abstract class ShapeValues<T extends ToXContentFragment> {
```

Comment:
Do we need this class in this PR?

Reply:
Done in e52ba263c3d118caf23d1da4358b9f8d95076702

Reply:
Good catch, indeed. I originally pulled in some of the other generic refactoring from the prototype, realised I did not need it and reverted the classes that extend this class, but forgot to remove this class itself!